### DISCUSSION TITLE
How to add script to unlisted page

### DISCUSSION THREAD

**Discussion by ming535** [wxt-dev/wxt#633] [General]
Hi, I am adding an "offscreen" page to my extension, and I am using "Unlisted Pages" and "Unlisted Scripts". (I am actually not sure if this is the right way to do it).

I put `offscreen.html` inside `entripoints`. Creating offscreen from existing html is working fine:
```javascript
    const offscreenUrl = browser.runtime.getURL("/offscreen.html");
    await chrome.offscreen
      .createDocument({
        url: offscreenUrl,
        reasons: [chrome.offscreen.Reason.USER_MEDIA],
        justification: "some test",
      })
```

I would like to load a scrip in `offscreen.html`, but it seems doesn't work. Here is my `offscreen.html`:

```html
<!doctype html>
<script src="./offscreen-main.ts" type="module"></script>
```

My `offscreen-main.ts` looks like this:
```
export default defineUnlistedScript({
  main() {
    console.log("offscreen run");
    throw new Error("offscreen run error");
  },
});
```

But I cannot see any log or exception in the offscreen page. Is there anything I am doing wrong?

**Comment by ming535** [comment]
I am able to run it by doing this:

```javascript
const def = defineUnlistedScript(() => {
  console.log("offscreen run");
});

console.log("hi offscreen ");
def.main();

export default def;
```

**Comment by aklinker1** [comment]
You can just run your code directly, no need for `export default` or `defineUnlistedScript`. Those are only required for standalone scripts not loaded by HTML files.

You'll need to structure your files like this:

```
entrypoints/
  offscreen/
    index.html
    main.ts
```

Refer to the popup in the template as an example:

https://github.com/wxt-dev/wxt/tree/main/templates%2Fvanilla%2Fentrypoints%2Fpopup

**Comment by aklinker1** [comment]
You can just run your code directly, no need for `export default` or `defineUnlistedScript`. Those are only required for standalone scripts not loaded by HTML files.

You'll need to structure your files like this:

```
entrypoints/
  offscreen/
    index.html
    main.ts
```

Refer to the popup in the template as an example:

https://github.com/wxt-dev/wxt/tree/main/templates%2Fvanilla%2Fentrypoints%2Fpopup

**Comment by chalapathi-ct** [comment]
Is there any WXT core messaging for the offscreen document.

**Comment by chalapathi-ct** [comment]
I am getting error when I am doing,
This is my file structure 

entrypoints/
- offscreen/
  - main.ts
  - offscreen.html
- background.ts

**offscreen.html**

`<!DOCTYPE html>
<html>
  <head>
    <title>Offscreen</title>
  </head>
  <body>
    <script type="module" src="./main.ts"></script>
  </body>
</html>

**main.ts**
`
console.log("hello");
`

**background.ts**

`
import offscreenHtml from './offscreen/offscreen.html?url';

const onClick = async (tab: Tabs.Tab) => {
  const offscreenUrl = browser.runtime.getURL(offscreenHtml);
  browser.offscreen.createDocument({
    url: offscreenUrl,
    reasons: ["USER_MEDIA", "DISPLAY_MEDIA"],
    justification: "Recording from browser.tabCapture API",
  });
}


export default defineBackground({
  type: "module",
  main() {
    browser.action.onClicked.addListener(onClick);
  }
});
`

I am getting

** Failed to load resource: net::ERR_FILE_NOT_FOUND main.ts:1**

**Reply by aklinker1** [reply]
Use this to get the URL instead:

```ts
const offscreenUrl = browser.runtime.getURL("/offscreen.html");
```

**Reply by aklinker1** [reply]
You're trying to pass in a local file path, but that's not how `browser.runtime.url` works.

```ts
import offscreenHtml from './offscreen/offscreen.html?url';
```

It's a little confusing when you use a bundler like WXT, but you need to pass in paths relative to the _output_ directory, not a local file path.
